clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java

```diff
@@ -153,7 +154,16 @@ void runOnce() {
                 .map(rm -> rm.maximumTimeToWait(currentTimeMs))
                 .reduce(Long.MAX_VALUE, Math::min);
 
-        reapExpiredApplicationEvents(currentTimeMs);
+        List<CompletableEvent<?>> completableEvents = reapExpiredApplicationEvents(currentTimeMs);
+        
+        // If there is a metadata error, complete the completable events which are not passed by error event with the metadata error
+        if (networkClientDelegate.metadataError().isPresent()) {
+            Throwable metadataError = networkClientDelegate.metadataError().get();
+            completableEvents.stream()
+                    .filter(event -> !(event instanceof CompletableApplicationEvent && ((CompletableApplicationEvent<?>) event).isPassedByErrorEvent()))
+                    .forEach(event -> event.future().completeExceptionally(metadataError));
+            networkClientDelegate.clearMetadataError();
```

Comment:
should we encapsulate this in a kind of `maybeFailOnMetadataError` that would take the events? 

Then I wonder if we need to do the same also before actually processing each event?? I'm thinking that, events that do require subscription metadata (ex. CheckAndUpdatePositions), may be processed, not generate any request and complete right away, so they will never know about the metadata error right?  Right before:
https://github.com/apache/kafka/blob/57c4c386796028dc4144dfb50a2a786dac0a51a3/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java#L181

Do we need to `maybeFailOnMetadataError(event)` to cover this gap?

Note that it's not a replacement for the maybeFailOnMetadataError of awaiting events, I wonder if we need them both (to cover events that complete right away and never make it to the "awaiting" list, and also the ones that are left waiting).

Reply:
I don't think it's necessary to execute `maybeFailOnMetadataError` before `applicationEventProcessor.process(event)`. Because, under normal situation, metadata errors only occur during the processing of events. Therefore, checking for metadata errors before processing is unnecessary.
WDYT?

Reply:
the trick is that events may complete right away on the call to process, so they will never fail with the metadata error because we would call completeExceptionally on a future that is already complete right?  (I wonder if this gap is the reason why you still need the `ErrorEvent` for metadata? if we remove the error event propagation on ln 156, and fill this gap, would it work? it would be nice because there would be one single consistent way of propagating metadata errors: via the calling api event. Ex. we would get TopicAuthException on poll and on positions in the same way, just because the CheckAndUpdatePositions fails. Makes sense? just suggesting to try it out, I could still be missing something but can see the gap in how we completeExceptionally with metadta errors.

Reply:
> so they will never fail with the metadata error because we would call completeExceptionally on a future that is already complete right?

That makes sense. We could encounter a metadata error while processing the last event, and we should use that error to complete all application events at the beginning of the next loop.